module/zfs/metaslab.c

```diff
@@ -1422,6 +1389,265 @@ metaslab_ops_t *zfs_metaslab_ops = &metaslab_ndf_ops;
  * ==========================================================================
  */
 
+/*
+ * Wait for any in-progress metaslab loads to complete.
+ */
+void
+metaslab_load_wait(metaslab_t *msp)
+{
+	ASSERT(MUTEX_HELD(&msp->ms_lock));
+
+	while (msp->ms_loading) {
+		ASSERT(!msp->ms_loaded);
+		cv_wait(&msp->ms_load_cv, &msp->ms_lock);
+	}
+}
+
+/*
+ * Wait for any in-progress flushing to complete.
+ */
+void
+metaslab_flush_wait(metaslab_t *msp)
+{
+	ASSERT(MUTEX_HELD(&msp->ms_lock));
+
+	while (msp->ms_flushing)
+		cv_wait(&msp->ms_flush_cv, &msp->ms_lock);
+}
+
+/*
+ * This callback is used only for debug purposes and it assumes that
+ * the entry that it is applied to is not obsolete for the metaslab given.
+ */
+static int
+metaslab_load_unflushed_trees_cb(space_map_entry_t *sme, void *arg)
+{
+	metaslab_t *target_ms = arg;
+	vdev_t *target_vd = target_ms->ms_group->mg_vd;
+
+	uint64_t offset = sme->sme_offset;
+	uint64_t size = sme->sme_run;
+
+	if (target_vd->vdev_id != sme->sme_vdev)
+		return (0);
+
+	metaslab_t *ms =
+	    target_vd->vdev_ms[offset >> target_vd->vdev_ms_shift];
+	if (ms != target_ms)
+		return (0);
+
+	switch (sme->sme_type) {
+		case SM_ALLOC:
+			range_tree_remove_xor_add_segment(offset, offset + size,
+			    ms->ms_unflushed_frees, ms->ms_unflushed_allocs);
+			break;
+		case SM_FREE:
+			range_tree_remove_xor_add_segment(offset, offset + size,
+			    ms->ms_unflushed_allocs, ms->ms_unflushed_frees);
+			break;
+		default:
+			panic("invalid maptype_t");
+			break;
+	}
+	return (0);
+}
+
+/*
+ * This function is used for verification only as it is very inefficient
+ * (it goes to disk reading all the log space maps discarding most data read).
```

Comment:
Only a grammatical nit: please add a comma after "maps," as I keep reading that as "space maps which discard most data read" :).

Reply:
hah thanks for pointing that out

Reply:
huh! Now I do too. Thanks for pointing that out!